Foundation/NSString.swift

```diff
@@ -1423,10 +1423,20 @@ extension NSMutableString {
 
         if let findResults = CFStringCreateArrayWithFindResults(kCFAllocatorSystemDefault, _cfObject, target._cfObject, CFRange(searchRange), options._cfValue(true)) {
             let numOccurrences = CFArrayGetCount(findResults)
+            let start = _storage.utf16.startIndex
+            var newStorage = Substring()
+            var sourceStringCurrentIndex = _storage.startIndex
             for cnt in 0..<numOccurrences {
-                let rangePtr = CFArrayGetValueAtIndex(findResults, backwards ? cnt : numOccurrences - cnt - 1)
-                replaceCharacters(in: NSRange(rangePtr!.load(as: CFRange.self)), with: replacement)
```

Comment:
removing this changes the behavior for subclassing

Reply:
Currently the concrete class is NsMutableString itself so ‘type(of: self) === NSMutableString.self’ would be true for the concrete case but not the subclass case

Reply:
Checking the class seems like a reasonable workaround to me.

Reply:
Can you use something like the following but specialise the `substring(with:)` for `NSMutableString` to avoid the `String()` copy it uses:

```
var newStorage = ""
var currentIndex = 0
for cnt in 0..<numOccurrences {
    let rangePtr = CFArrayGetValueAtIndex(findResults, backwards ? numOccurrences - cnt - 1 : cnt)
    let range = NSRange(rangePtr!.load(as: CFRange.self))
    let copyRange = NSRange(location: currentIndex, length: range.location)
    newStorage += substring(with: copyRange)
    newStorage += replacement
    currentIndex = range.location + range.length
}
let copyRange = NSRange(location: currentIndex, length: length - currentIndex)
newStorage += substring(with: copyRange)
setString(newStorage)
```